group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -536,6 +538,27 @@ public OffsetFetchResponseData.OffsetFetchResponseGroup fetchAllOffsets(
             .setTopics(topicResponses);
     }
 
+    public void cleanupOffsetMetadata(List<Record> records, long offsetsRetentionMs) {
+        // TODO: get only committed offset groups?
+        offsetsByGroup.forEach((groupId, offsetsByTopic) -> {
+            try {
+                Group group = groupMetadataManager.group(groupId);
+                Map<TopicPartition, OffsetAndMetadata> expiredOffsets = group.expiredOffsets(offsetsRetentionMs, offsetsByTopic);
+                log.debug("[GroupId {}] Expiring offsets: {}", groupId, expiredOffsets.keySet());
+                expiredOffsets.forEach((topicPartition, offsetAndMetadata) -> {
```

Comment:
Is expiredOffsets really necessary? It seems that we could directly generate the required records instead of building it, no?